ddtrace/internal/writer.py

```diff
@@ -64,19 +67,41 @@ def flush_queue(self):
             log.error('error while filtering traces: {0}'.format(err))
             return
 
-        if not traces:
-            return
+        traces_flushed = len(traces)
+        traces_flushed_error = 0
 
         # If we have data, let's try to send it.
         traces_responses = self.api.send_traces(traces)
         for response in traces_responses:
             if isinstance(response, Exception) or response.status >= 400:
                 self._log_error_status(response)
+                traces_flushed_error += 1
             elif self._priority_sampler:
                 result_traces_json = response.get_json()
                 if result_traces_json and 'rate_by_service' in result_traces_json:
                     self._priority_sampler.set_sample_rate_by_service(result_traces_json['rate_by_service'])
 
+        # Dump statistics
+        # NOTE: Do not use the buffering of dogstatsd as it's not thread-safe
+        # https://github.com/DataDog/datadogpy/issues/439
+        if self.dogstatsd:
+            self.dogstatsd.increment('datadog.tracer.traces.api.requests', len(traces_responses))
+            for status, grouped_responses in itertools.groupby(
+                    sorted(traces_responses, key=lambda r: r.status),
+                    key=lambda r: r.status):
+                self.dogstatsd.increment('datadog.tracer.traces.api.responses.count',
+                                         len(list(grouped_responses)),
+                                         tags=['status:%d' % status])
+            if hasattr(time, 'thread_time_ns'):
+                self.dogstatsd.increment('datadog.tracer.writer.cpu_time', time.thread_time_ns())
+            self.dogstatsd.increment('datadog.tracer.traces.flushed', traces_flushed)
+            self.dogstatsd.increment('datadog.tracer.traces.flushed.errors', traces_flushed_error)
+            dropped, enqueued, enqueued_lengths = self._trace_queue.reset_stats()
+            self.dogstatsd.increment('datadog.tracer.traces.dropped', dropped)
```

Comment:
hmm, re-thinking these ones a bit, I wonder if we need to have `enqueued` be `total - dropped`, since right now it looks like `enqueued` includes those that were dropped as well.

Same for the `enqueued_lengths`, that is the length of traces we called `.put()` on, not that were enqueued to be flushed.

Reply:
Enqueued is interesting because its derive represents the rate at which you actually put traces in the queue. We need that. Same for the length, we need to know the (average) length of traces that are put in queue to have an idea of the workload that is generated.

I don't know what `total` mean in your example so I'm not sure what you think we miss here?

Reply:
there is, the total number of times we called `put`, the total number of traces that we "kept", the total number of traces we "dropped"

`enqueued` is measuring calls to `put`, which is fine, but want to make sure we don't ever confuse that with the total number of traces kept for flushing.

Reply:
Yes; I've a new version of the patch that depends on #1071 that is a bit clearer I think. Will post it once the latter is merged.